PBL-32068 Remove dist folder and let travis manage it on gh-pages

HBehrens opened the pull request:
- creates new grunt task `build-missing-diets` to update `build/dist` based on remote tags
- configures travis to call this each time (as part of a new `pre-publish` task) it rebuilds the static page
- remove dist folder and change grunt tasks to directly build into `build/dist`

I also introduced a new `tasks` folder that contains this new task + a few convenience functions to deal with promises. The idea is to continue to use this for future tasks and also to clean up today's `Grunfile.js` whenever we find the time.

HBehrens commented:
@sirlantis @franc0is I think this should cover our goals, can you please review?

I used promises to avoid as much of the callback hell as possible. Switching to ES2015 + yield could make it more readable but I'd like avoid this step at this point.

sirlantis commented on package.json:
```diff
@@ -35,6 +35,8 @@
         "load-grunt-tasks": "^3.4.0",
-        "mocha": "^2.3.4"
+        "mocha": "^2.3.4",
+        "q": "^1.4.1",
```
Just a hint towards https://www.npmjs.com/package/native-or-bluebird here

HBehrens replied:
Instead of putting work here, I created #31 to move to an up-to-date version of Node directly.

sirlantis commented on tasks/dist.js:
```diff
@@ -0,0 +1,106 @@
+module.exports = function(grunt) {
+
+    // location of our temporary working copy we use to build tags
+    var tempCheckout = 'temp/checkout';
```
(unimportant) The more common abbreviation for 'temporary' is `tmp`.

HBehrens replied:
:white_check_mark:

sirlantis commented on tasks/dist.js:
```diff
@@ -0,0 +1,106 @@
+            return exec('npm install', {cwd: tempCheckout});
+        })
+        .then(function() {
+            return exec('grunt build', {cwd:tempCheckout});
```
(unimportant) space missing behind `cwd:`

HBehrens replied:
:white_check_mark:

sirlantis commented on tasks/dist.js:
```diff
@@ -0,0 +1,106 @@
+                grunt.log.warn("skipping file", destName, "Tag", tag, "seems to be named incorrectly.");
+                return false;
+            }
+            return exec('cp ' + tempCheckout + '/' + destName + ' ' + destName);
```
string concatenation with remote inputs (git tags) to build commands scare me

HBehrens replied:
These tags are following the aforementioned pattern and should be secure. Will add a comment about it, let me know if you see any other problem.

sirlantis commented on tasks/dist.js:
```diff
@@ -0,0 +1,106 @@
+    var tempCheckout = 'temp/checkout';
+
+    // register a few promise related convenience functions
+    var exec = require('./utils')(grunt).exec;
```
hint at https://github.com/kevinbeaty/fs-promise

HBehrens replied:
I would still need to create a chained promise to keep the logging aspect. Is replacing a sync'ed call worth the dependency?

sirlantis replied:
You could at least also remove the `rm -rf` and replace one `cp`. I would avoid shelling out as much as possible (the only cases where you have to is when you call grunt again or want to use the globbing).

sirlantis commented on tasks/dist.js:
```diff
@@ -0,0 +1,106 @@
+        // filter tags: keep version tags we don't have a rocky-x.y.z.js for, yet
+        .then(function(tags) {
+                var foundTags = tags.all.filter(function(tag) {
+                    return tag.match(/^v?\d\.\d\.\d$/);
```
Should be `/^v?\d+\.\d+\.\d+$/`

HBehrens replied:
:white_check_mark:

sirlantis commented on tasks/dist.js:
```diff
@@ -0,0 +1,106 @@
+            });
+            return tasks.reduce(function(chain, t){
+                return chain.then(t);
+            }, Q(true));
```
JFYI: `return tasks.reduce(Q.when, Q(true));` does the same as these three lines

HBehrens replied:
:white_check_mark:

sirlantis commented on tasks/dist.js:
```diff
@@ -0,0 +1,106 @@
+            throw error;
+        })
+        // correctly terminate async grunt task
+        .done(function(){done(true)}, function(){done(false)});
```
I would merge the log from the `catch` above into the error handler here and remove the `catch`

HBehrens replied:
:white_check_mark:

sirlantis commented on tasks/utils.js:
```diff
@@ -0,0 +1,54 @@
+        grunt.log.write(logMessage);
+        var deferred = Q.defer();
+        var args = Array.prototype.slice.call(arguments, 1);
+        args.push(function(err, data) {
```
Since you go full on with the Q dependency, this function can probably be simplified by just using (logging to be added):

```
var promisified = Q.denodeify(this[fnName].bind(this));
return promisified(Array.prototype.slice.call(arguments, 1));
```

HBehrens replied:
I changed it to `Q.nfapply(this[fnName].bind(this), Array.prototype.slice.call(arguments, 1))`

sirlantis commented on tasks/utils.js:
```diff
@@ -0,0 +1,54 @@
+    // promise wrapper around exec + logging
+    var _exec = Q.nfbind(require('child_process').exec);
+    var exec = function(cmd, options) {
+        grunt.log.write("exec:", cmd, options ? options : {}, "...");
```
`options || {}`

HBehrens replied:
:white_check_mark: incredible how quickly one forgets idioms like these…

sirlantis commented on tasks/dist.js:
```diff
@@ -0,0 +1,106 @@
+            // copy all rocky.js files from gh-pages' dist into this working copy's build folder, but
+            // preserve existing files in build/dist (presumably only the one we built from this working copy before)
+            return exec('cp -n ' + tempCheckout + '/dist/*.js build/dist')
+                   .catch(function(){return "this is fine since cp -n returns 1 if there was a file already."});
```
Instead of swallowing an error here (`cp` can fail because of many reasons), why not remove build/dist before?

sirlantis replied:
Or maybe really just add the dependency to `glob` and do it properly.

var glob = Q.nfbind(require('glob'))

glob(path.join(tempCheckout, 'dist', '*.js')
  .then(function(sources) {
    Q.all(sources.map(function(source) {
      var dest = path.join('build', 'dist');
      return Q.nfbind(fs.exists)(dest)
      .then(function(exists) {
        if (!exists) {
          return Q.nfbind(fs.copy)(src, 'build/dist')
        }
      }
    });
  });

HBehrens replied:
This looks terrible! What about this: I keep it as it is and you create a PR that bumps the node version and replaces this cumbersome snippets with `yield` + promises?

HBehrens replied:
So, the new code only copies the single resulting file from `package.json#main` and does so without `exec('cp…')` :white_check_mark:

sirlantis commented on tasks/dist.js:
```diff
@@ -0,0 +1,117 @@
+
+    // checks for a valid version string to find tags and verify valid filed names, e.g. x.y.z, x.y.z-beta1
+    function isValidRockyVersion(version) {
+        return version.match(/^\d+\.\d+\.\d+$(-\n+)?/);
```
I think the `$` has to go behind the suffix ;)

HBehrens replied:
oops :white_check_mark:

franc0is commented:
This should also create a "rocky-stable.js" when it finds a tag for a version newer than the newest file in dist. Or do you want that to happen in a follow on PR?

franc0is commented on tasks/dist.js:
```diff
@@ -0,0 +1,117 @@
+    grunt.registerTask(taskName, desc, function() {
+        var done = this.async();
+
+        // delete temp folder
```
How about using https://www.npmjs.com/package/tmp?

HBehrens replied:
I switched it to `.grunt/<task_name>` as the other tasks did already.

HBehrens commented:
@franc0is can you have a final look before I merge this?

HBehrens commented:
I also removed the test tags/branches and replaced `v0.1.0` with `0.1.0`. Going forward, master will carry the most-recent version in its `package.json` to simplify local npm-based builds and we bump it as part of the (yet to be automated) release process.

franc0is commented:
:+1:
